core/capabilities/triggers/logevent/trigger.go

```diff
@@ -0,0 +1,178 @@
+	}
+
+	// Create a New Contract Reader client, which brings a corresponding ContractReader gRPC service
+	// in Chainlink Core service
+	contractReader, err := relayer.NewContractReader(ctx, jsonBytes)
+	if err != nil {
+		return nil, nil,
+			fmt.Errorf("error fetching contractReader for chainID %d from relayerSet: %v", logEventConfig.ChainId, err)
+	}
+
+	// Bind Contract in ContractReader
+	boundContracts := []types.BoundContract{{Name: reqConfig.ContractName, Address: reqConfig.ContractAddress.Hex()}}
+	err = contractReader.Bind(ctx, boundContracts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Get current block HEAD/tip of the blockchain to start polling from
+	latestHead, err := relayer.LatestHead(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error getting latestHead from relayer client: %v", err)
+	}
+	height, err := strconv.ParseUint(latestHead.Height, 10, 64)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid height in latestHead from relayer client: %v", err)
+	}
+
+	// Setup callback channel, logger and ticker to poll ContractReader
+	callbackCh := make(chan capabilities.TriggerResponse, defaultSendChannelBufferSize)
+	ticker := time.NewTicker(time.Duration(logEventConfig.PollPeriod) * time.Millisecond)
+	done := make(chan bool)
+	lggr, err := logger.New()
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not initialise logger for LogEventTrigger")
+	}
+
+	// Initialise a Log Event Trigger
+	l := &logEventTrigger{
+		ch:   callbackCh,
+		lggr: logger.Named(lggr, "LogEventTrigger: "),
+		ctx:  ctx,
+
+		reqConfig:      reqConfig,
+		contractReader: contractReader,
+		relayer:        relayer,
+		startBlockNum:  height,
+
+		logEventConfig: logEventConfig,
+		ticker:         ticker,
+		done:           done,
+	}
+	go l.Listen()
+
+	return l, callbackCh, nil
+}
+
+// Listen to contract events and trigger workflow runs
+func (l *logEventTrigger) Listen() {
+	// Listen for events from lookbackPeriod
+	var logs []types.Sequence
+	var err error
+	logData := make(map[string]any)
+	limitAndSort := query.LimitAndSort{
+		SortBy: []query.SortBy{query.NewSortByTimestamp(query.Asc)},
+	}
+	for {
+		select {
+		case <-l.done:
+			return
+		case t := <-l.ticker.C:
+			l.lggr.Infof("Polling event logs from ContractReader using QueryKey at", t)
+			logs, err = l.contractReader.QueryKey(
+				l.ctx,
+				types.BoundContract{Name: l.reqConfig.ContractName, Address: l.reqConfig.ContractAddress.Hex()},
+				query.KeyFilter{
+					Key: l.reqConfig.ContractEventName,
+					Expressions: []query.Expression{
+						query.Confidence(primitives.Finalized),
+						query.Block(fmt.Sprintf("%d", l.startBlockNum-l.logEventConfig.LookbackBlocks), primitives.Gte),
+					},
+				},
+				limitAndSort,
+				logData,
+			)
+			if err != nil {
+				l.lggr.Fatalw("QueryKey failure", "err", err)
+				continue
+			}
+			for _, log := range logs {
+				triggerResp := createTriggerResponse(log)
+				go func(resp capabilities.TriggerResponse) {
```

Comment:
What if there's an error sending to channel?

Reply:
Having the channel open whenever you are trying to write is an invariant that you must maintain.

These go routines need to be tracked too (e.g. with a wait group that blocks the parent from returning until they are complete). They can't be orphaned in the background, disconnected from the rest of the system.

Reply:
Sure, I've added a waitGroup and waiting for it in Stop